### ISSUE TITLE
Why print so many relu layers？

### ISSUE THREAD

**Issue by Zhengyu1998** [TylerYep/torchinfo#105]
Why print so many relu layers？

This is my net
```
    def __init__(self, input_nc=4, output_nc=2, ngf=64, n_downsampling=3, n_blocks=9, norm_layer=nn.BatchNorm2d, padding_type='reflect'):

        assert(n_blocks >= 0)
        super().__init__()

        self.training = True
        self.input_nc = input_nc
        self.output_nc = output_nc

        activation = nn.ReLU(True) 

        model = [nn.ReflectionPad2d(3), nn.Conv2d(input_nc, ngf, kernel_size=7, padding=0), norm_layer(ngf), activation]

        ### downsample
        for i in range(n_downsampling):
            mult = 2**i
            model += [nn.Conv2d(ngf * mult, ngf * mult * 2, kernel_size=3, stride=2, padding=1),
                      norm_layer(ngf * mult * 2), activation]

        ### resnet blocks
        mult = 2**n_downsampling
        for i in range(n_blocks):
            model += [ResnetBlock(ngf * mult, padding_type=padding_type, activation=activation, norm_layer=norm_layer)]   

        ### upsample         
        for i in range(n_downsampling):
            mult = 2**(n_downsampling - i)
            model += [nn.ConvTranspose2d(ngf * mult, int(ngf * mult / 2), kernel_size=3, stride=2, padding=1, output_padding=1),
                       norm_layer(int(ngf * mult / 2)), activation]
        model += [nn.ReflectionPad2d(3), nn.Conv2d(ngf, output_nc, kernel_size=7, padding=0)]
        print(model)        
        self.model = nn.Sequential(*model) 
```

the summary print many relu layers like this screenshot
![Screenshot from 2022-01-04 22-37-34](https://user-images.githubusercontent.com/49787797/148075371-a111c383-5326-4169-aaff-ed18d50606ed.png)

**Comment by TylerYep** [comment]
Can you post your full code? This looks like a bug.

**Comment by Zhengyu1998** [comment]
@TylerYep  OK
**This is my net**
```
class TestNet(nn.Module):
    def __init__(self, input_nc=4, output_nc=2, ngf=64, n_downsampling=3, n_blocks=9, norm_layer=nn.BatchNorm2d, padding_type='reflect'):

        assert(n_blocks >= 0)
        super().__init__()

        self.training = True
        self.input_nc = input_nc
        self.output_nc = output_nc

        activation = nn.ReLU(True) 

        model = [nn.ReflectionPad2d(3), nn.Conv2d(input_nc, ngf, kernel_size=7, padding=0), norm_layer(ngf), activation]

        ### downsample
        for i in range(n_downsampling):
            mult = 2**i
            model += [nn.Conv2d(ngf * mult, ngf * mult * 2, kernel_size=3, stride=2, padding=1),
                      norm_layer(ngf * mult * 2), activation]

        ### resnet blocks
        mult = 2**n_downsampling
        for i in range(n_blocks):
            model += [ResnetBlock(ngf * mult, padding_type=padding_type, activation=activation, norm_layer=norm_layer)]   

        ### upsample         
        for i in range(n_downsampling):
            mult = 2**(n_downsampling - i)
            model += [nn.ConvTranspose2d(ngf * mult, int(ngf * mult / 2), kernel_size=3, stride=2, padding=1, output_padding=1),
                       norm_layer(int(ngf * mult / 2)), activation]
        model += [nn.ReflectionPad2d(3), nn.Conv2d(ngf, output_nc, kernel_size=7, padding=0)]
        print(model)        
        self.model = nn.Sequential(*model)    

    def forward(self, input):
        return self.model(input)   
```

**This is the ResNetBlock**
```
class ResnetBlock(nn.Module):
    def __init__(self, dim, padding_type, norm_layer, activation=nn.ReLU(True), use_dropout=False):
        super(ResnetBlock, self).__init__()
        self.conv_block = self.build_conv_block(dim, padding_type, norm_layer, activation, use_dropout)

    def build_conv_block(self, dim, padding_type, norm_layer, activation, use_dropout):
        conv_block = []
        p = 0
        if padding_type == 'reflect':
            conv_block += [nn.ReflectionPad2d(1)]
        elif padding_type == 'replicate':
            conv_block += [nn.ReplicationPad2d(1)]
        elif padding_type == 'zero':
            p = 1
        else:
            raise NotImplementedError('padding [%s] is not implemented' % padding_type)

        conv_block += [nn.Conv2d(dim, dim, kernel_size=3, padding=p),
                       norm_layer(dim),
                       activation]
        if use_dropout:
            conv_block += [nn.Dropout(0.5)]

        p = 0
        if padding_type == 'reflect':
            conv_block += [nn.ReflectionPad2d(1)]
        elif padding_type == 'replicate':
            conv_block += [nn.ReplicationPad2d(1)]
        elif padding_type == 'zero':
            p = 1
        else:
            raise NotImplementedError('padding [%s] is not implemented' % padding_type)
        conv_block += [nn.Conv2d(dim, dim, kernel_size=3, padding=p),
                       norm_layer(dim)]

        return nn.Sequential(*conv_block)

    def forward(self, x):
        out = x + self.conv_block(x)
        return out   
```
**I call summary() as follows**
```
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
net = TestNet()
net = net.to(device)
summary(net, (4, 4, 512, 512))
```

**Comment by FrancescoSaverioZuppichini** [comment]
I know why! It happened to me at some point when I was less wise :) 

So basically you have a default parameter 

`activation=nn.ReLU(True),`

function default parameters are evaluated at module time, not running time so your activation is always the same one, the same reference. Example


```python

def function(a=[]):
     a.append(1)
     return a

function() # a = [1]
function() # a = [1,1]
```

**Comment by Zhengyu1998** [comment]
@TylerYep 
Ohhhhh, thank you so much!

**Comment by TylerYep** [comment]
@FrancescoSaverioZuppichini  Great catch!! Really appreciate your help here. An excellent example of torchinfo surfacing unintentional behavior in your models at runtime that `print(model)` wouldn't find.  :)

**Comment by TylerYep** [comment]
But I still see an actual issue with torchinfo related to this code, reproduced here:
```

def test_too_many_linear() -> None:
    class TestNet(nn.Module):
        def __init__(self, n_downsampling=3):
            super().__init__()
            activation = nn.Linear(50, 50)
            model = [activation, nn.ReLU(True)]
            for _ in range(n_downsampling):
                model += [activation, nn.ReLU(True)]
            self.model = nn.Sequential(*model)

        def forward(self, x):
            return self.model(x)

    summary(TestNet(), (4, 50))


def test_too_many_relus() -> None:
    class TestNet(nn.Module):
        def __init__(self, ngf=64, n_downsampling=3):
            super().__init__()
            activation = nn.ReLU(True)
            model = [
                nn.ReflectionPad2d(3),
                nn.Conv2d(4, ngf, kernel_size=7, padding=0),
                nn.BatchNorm2d(ngf),
                activation,
            ]
            for i in range(n_downsampling):
                mult = 2 ** i
                model += [
                    nn.Conv2d(
                        ngf * mult, ngf * mult * 2, kernel_size=3, stride=2, padding=1
                    ),
                    nn.BatchNorm2d(ngf * mult * 2),
                    activation,
                ]
            self.model = nn.Sequential(*model)

        def forward(self, x):
            return self.model(x)

    summary(TestNet(), (4, 4, 512, 512))
```

I'll leave this issue open to continue investigating and post once I have a fix!

**Comment by TylerYep** [comment]
This bug has been fixed in torchinfo v1.6.2.

Your model should now output:
```
===============================================================================================
Layer (type:depth-idx)                        Output Shape              Param #
===============================================================================================
TestNet                                       --                        --
├─Sequential: 1-1                             [4, 2, 512, 512]          --
│    └─ReflectionPad2d: 2-1                   [4, 4, 518, 518]          --
│    └─Conv2d: 2-2                            [4, 64, 512, 512]         12,608
│    └─BatchNorm2d: 2-3                       [4, 64, 512, 512]         128
│    └─ReLU: 2-4                              [4, 64, 512, 512]         --
│    └─Conv2d: 2-5                            [4, 128, 256, 256]        73,856
│    └─BatchNorm2d: 2-6                       [4, 128, 256, 256]        256
│    └─ReLU: 2-7                              [4, 128, 256, 256]        --
│    └─Conv2d: 2-8                            [4, 256, 128, 128]        295,168
│    └─BatchNorm2d: 2-9                       [4, 256, 128, 128]        512
│    └─ReLU: 2-10                             [4, 256, 128, 128]        --
│    └─Conv2d: 2-11                           [4, 512, 64, 64]          1,180,160
│    └─BatchNorm2d: 2-12                      [4, 512, 64, 64]          1,024
│    └─ReLU: 2-13                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-14                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-1                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-15                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-16                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-2                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-17                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-18                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-3                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-19                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-20                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-4                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-21                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-22                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-5                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-23                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-24                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-6                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-25                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-26                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-7                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-27                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-28                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-8                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-29                             [4, 512, 64, 64]          --
│    └─ResnetBlock: 2-30                      [4, 512, 64, 64]          --
│    │    └─Sequential: 3-9                   [4, 512, 64, 64]          4,721,664
│    └─ReLU: 2-31                             [4, 512, 64, 64]          --
│    └─ConvTranspose2d: 2-32                  [4, 256, 128, 128]        1,179,904
│    └─BatchNorm2d: 2-33                      [4, 256, 128, 128]        512
│    └─ReLU: 2-34                             [4, 256, 128, 128]        --
│    └─ConvTranspose2d: 2-35                  [4, 128, 256, 256]        295,040
│    └─BatchNorm2d: 2-36                      [4, 128, 256, 256]        256
│    └─ReLU: 2-37                             [4, 128, 256, 256]        --
│    └─ConvTranspose2d: 2-38                  [4, 64, 512, 512]         73,792
│    └─BatchNorm2d: 2-39                      [4, 64, 512, 512]         128
│    └─ReLU: 2-40                             [4, 64, 512, 512]         --
│    └─ReflectionPad2d: 2-41                  [4, 64, 518, 518]         --
│    └─Conv2d: 2-42                           [4, 2, 512, 512]          6,274
===============================================================================================
Total params: 45,614,594
Trainable params: 45,614,594
Non-trainable params: 0
Total mult-adds (T): 1.01
===============================================================================================
Input size (MB): 16.78
Forward/backward pass size (MB): 6325.01
Params size (MB): 182.46
Estimated Total Size (MB): 6524.25
===============================================================================================
```

Thank you for the issue!
